### PR TITLE
Offline map, geocode, and route demo from Dev Summit Europe

### PR DESCRIPTION
**Pull request by shellygill**
@doneill A PR for a demo used at the Dev Summit Europe last week. The app will not successfully run without data in place, but unfortunately we dont have distributable datasets for this. So the code is provided for info purposes only to show how you can use an MMPK to show a map (along with a TPK as basemap), and also to geocode (using merged locators) and solve a route.

### CODE REVIEW

File: 2016-EDS/GeocodeAndRoute/app/src/main/java/com/esri/arcgisruntime/demo/geocodeandroute/GeocodeRouteActivity.java [hunk 1]
```diff
@@ -0,0 +1,615 @@
...
+      Snackbar.make(mMapView, getString(R.string.hydrant_stop_not_set), Snackbar.LENGTH_SHORT).show();
+      return;
+    }
+
+    if (mRouteTask == null ) {
+      Snackbar.make(mMapView, getString(R.string.route_task_not_set), Snackbar.LENGTH_SHORT).show();
+      return;
+    }
+
+    RouteParameters routeParams = null;
```
Comment (doneill) [inline 2016-EDS/GeocodeAndRoute/app/src/main/java/com/esri/arcgisruntime/demo/geocodeandroute/GeocodeRouteActivity.java]: Initializing as null is redundant and can be removed

File: 2016-EDS/GeocodeAndRoute/app/src/main/java/com/esri/arcgisruntime/demo/geocodeandroute/GeocodeRouteActivity.java [hunk 2]
```diff
@@ -0,0 +1,615 @@
...
+      finish.setRouteName(getString(R.string.route_name));
+      finish.setName(getString(R.string.stop2_name));
+      routeParams.getStops().add(finish);
+
+      final ListenableFuture<RouteResult> routeFuture = mRouteTask.solveRouteAsync(routeParams);
+      routeFuture.addDoneListener(new Runnable() {
+        @Override
+        public void run() {
+          // Show results of solved route.
+          RouteResult routeResult = null;
```
Comment (doneill) [inline 2016-EDS/GeocodeAndRoute/app/src/main/java/com/esri/arcgisruntime/demo/geocodeandroute/GeocodeRouteActivity.java:531]: Initializing as null is redundant and can be removed

File: 2016-EDS/GeocodeAndRoute/app/src/main/java/com/esri/arcgisruntime/demo/geocodeandroute/GeocodeRouteActivity.java [hunk 3]
```diff
@@ -0,0 +1,615 @@
...
+import android.support.v4.content.ContextCompat;
+import android.text.TextUtils;
+import android.view.View;
+import android.support.design.widget.NavigationView;
+import android.support.v4.view.GravityCompat;
+import android.support.v4.widget.DrawerLayout;
+import android.support.v7.app.ActionBarDrawerToggle;
+import android.support.v7.app.AppCompatActivity;
+import android.support.v7.widget.Toolbar;
+import android.view.Menu;
```
Comment (doneill) [inline 2016-EDS/GeocodeAndRoute/app/src/main/java/com/esri/arcgisruntime/demo/geocodeandroute/GeocodeRouteActivity.java]: unused import

File: 2016-EDS/GeocodeAndRoute/app/src/main/java/com/esri/arcgisruntime/demo/geocodeandroute/GeocodeRouteActivity.java [hunk 4]
```diff
@@ -0,0 +1,615 @@
...
+import com.esri.arcgisruntime.tasks.networkanalysis.Stop;
+
+import java.io.File;
+import java.util.List;
+import java.util.concurrent.ExecutionException;
+
+public class GeocodeRouteActivity extends AppCompatActivity
+    implements NavigationView.OnNavigationItemSelectedListener {
+
+  private static final String TAG = GeocodeRouteActivity.class.getSimpleName();
```
Comment (doneill) [inline 2016-EDS/GeocodeAndRoute/app/src/main/java/com/esri/arcgisruntime/demo/geocodeandroute/GeocodeRouteActivity.java]: Unused declaration

File: 2016-EDS/GeocodeAndRoute/app/src/main/java/com/esri/arcgisruntime/demo/geocodeandroute/GeocodeRouteActivity.java [hunk 5]
```diff
@@ -0,0 +1,615 @@
...
+      @Override
+      public void onClick(View view) {
+        solveRoute();
+      }
+    });
+
+    DrawerLayout drawer = (DrawerLayout) findViewById(R.id.drawer_layout);
+    ActionBarDrawerToggle toggle = new ActionBarDrawerToggle(
+        this, drawer, toolbar, R.string.navigation_drawer_open, R.string.navigation_drawer_close);
+    drawer.setDrawerListener(toggle);
```
Comment (doneill) [inline 2016-EDS/GeocodeAndRoute/app/src/main/java/com/esri/arcgisruntime/demo/geocodeandroute/GeocodeRouteActivity.java]: `DrawerLayout.setDrawerListener()` is deprecated, use `addDrawerListener()`.

File: 2016-EDS/GeocodeAndRoute/Readme.md [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+![GeocodeAndRoute Screenshot](Screenshot_GeocodeAndRoute.png)
+
+## APIs used
+- `MobileMapPackage`
+- `ArcGISTiledLayer`
+- `LocatorTask`
+- `RouteTask`
+
+## Prerequisites
+- The data used in this demo is not available for distribution. The code is therefore provided for information only, and can be adapted to your own datasets. The app requires a mobile map package (.mmpk) containing at least one map, with a corresponding locator and network dataset combined with operational data in the map. A tile cache (.tpk) file is required to be used as the basemap layer for the map. The code assumes specific locator names in order to assign the geocoded location as either an address or a hydrant location.
```
Comment (doneill) [inline 2016-EDS/GeocodeAndRoute/Readme.md:16]: You should explicitly explain where to make edits in code to include your own mmpk and tpk files.  You should also note the viewpoint may need to be adjusted to your dataset.

**Review by doneill — COMMENTED** [review]
@shellygill Some minor comments for your review. I am still looking for some data to include with this as well.

File: 2016-EDS/GeocodeAndRoute/app/src/main/java/com/esri/arcgisruntime/demo/geocodeandroute/GeocodeRouteActivity.java (hunk 3 above)
Comment (shellygill) [reply 2016-EDS/GeocodeAndRoute/app/src/main/java/com/esri/arcgisruntime/demo/geocodeandroute/GeocodeRouteActivity.java]: Removed.

File: 2016-EDS/GeocodeAndRoute/app/src/main/java/com/esri/arcgisruntime/demo/geocodeandroute/GeocodeRouteActivity.java (hunk 4 above)
Comment (shellygill) [reply 2016-EDS/GeocodeAndRoute/app/src/main/java/com/esri/arcgisruntime/demo/geocodeandroute/GeocodeRouteActivity.java]: Removed.

File: 2016-EDS/GeocodeAndRoute/app/src/main/java/com/esri/arcgisruntime/demo/geocodeandroute/GeocodeRouteActivity.java (hunk 5 above)
Comment (shellygill) [reply 2016-EDS/GeocodeAndRoute/app/src/main/java/com/esri/arcgisruntime/demo/geocodeandroute/GeocodeRouteActivity.java]: Updated.

File: 2016-EDS/GeocodeAndRoute/app/src/main/java/com/esri/arcgisruntime/demo/geocodeandroute/GeocodeRouteActivity.java (hunk 1 above)
Comment (shellygill) [reply 2016-EDS/GeocodeAndRoute/app/src/main/java/com/esri/arcgisruntime/demo/geocodeandroute/GeocodeRouteActivity.java]: Updated.

File: 2016-EDS/GeocodeAndRoute/app/src/main/java/com/esri/arcgisruntime/demo/geocodeandroute/GeocodeRouteActivity.java (hunk 2 above)
Comment (shellygill) [reply 2016-EDS/GeocodeAndRoute/app/src/main/java/com/esri/arcgisruntime/demo/geocodeandroute/GeocodeRouteActivity.java:531]: Updated.

File: 2016-EDS/GeocodeAndRoute/Readme.md (hunk 6 above)
Comment (shellygill) [reply 2016-EDS/GeocodeAndRoute/Readme.md:16]: As its demo code, and was never written to be really generic, so there's more cases than these that you'd need to review in light of your own data. I've added comments to indicate lines that might be data-specific.

**Comment by shellygill** [comment]
Thanks @doneill back to you.

**Comment by doneill** [comment]
Looks good, will merge.

**Pull request merged**